src/OmniSharp/Workers/Refactoring/FixUsings.cs

```diff
@@ -0,0 +1,238 @@
+            await RemoveDuplicateUsings(fileName, workspace, document, semanticModel, codeActionProvider);
+            document = workspace.GetDocument(fileName);
+            var compilationUnitSyntax = (await AddLinqQuerySyntax(fileName, workspace, document, semanticModel));
+            document = document.WithSyntaxRoot(compilationUnitSyntax);
+            var docText = await document.GetTextAsync();
+
+            return new FixUsingsResponse(docText.ToString(), ambiguous);
+        }
+
+        private static async Task<List<QuickFix>> AddMissingUsings(string fileName, OmnisharpWorkspace workspace, Document document, SemanticModel semanticModel, ICodeActionProvider codeActionProvider)
+        {
+            bool processMore = true;
+            var ambiguousNodes = new List<SimpleNameSyntax>();
+            var ambiguous = new List<QuickFix>();
+
+            while (processMore)
+            {
+                bool updated = false;
+
+                var syntaxNode = (await document.GetSyntaxTreeAsync()).GetRoot();
+                var nodes = syntaxNode.DescendantNodes()
+                    .OfType<SimpleNameSyntax>()
+                    .Where(x => semanticModel.GetSymbolInfo(x).Symbol == null && !ambiguousNodes.Contains(x)).ToList();
+
+                foreach (var node in nodes)
+                {
+                    document = workspace.GetDocument(fileName);
+                    semanticModel = await document.GetSemanticModelAsync();
+                    var sourceText = (await document.GetTextAsync());
+                    var diagnostics = semanticModel.GetDiagnostics();
+                    var location = node.Identifier.Span;
+                    var actions = new List<CodeAction>();
+                    var usingOperations = new Dictionary<string, CodeActionOperation>();
+                    var pointDiagnostics = diagnostics.Where(d => d.Location.SourceSpan.Contains(location)).ToImmutableArray();
+
+                    if (pointDiagnostics.Any())
+                    {
+                        var pointdiagfirst = pointDiagnostics.First().Location.SourceSpan;
+                        if (pointDiagnostics.Any(d => d.Location.SourceSpan != pointdiagfirst))
+                        {
+                            continue;
+                        }
+                        var context = new CodeFixContext(document, pointDiagnostics.First().Location.SourceSpan, pointDiagnostics, (a, d) => actions.Add(a), CancellationToken.None);
+                        var providers = codeActionProvider.CodeFixes;
+
+                        foreach (var provider in providers)
+                        {
+                            var register = provider.RegisterCodeFixesAsync(context);
+
+                            foreach (var action in actions)
+                            {
+                                if (action.Title.StartsWith("using"))
```

Comment:
I think that instead of looping over the providers and then actions, you should create an instance of the CodeAction that you want to use (will probably need reflection) and keep it around between requests.